File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.java
import java.util.*;

class Solution {
    public int solution(String[] board) {
        int[][] offset = { {0, 1}, {0, -1}, {1, 0}, {-1, 0} };
        int rows = board.length;
        int cols = board[0].length();
        boolean[][] visited = new boolean[rows][cols];
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (board[i].charAt(j) == 'R') {
                    Queue<int[]> bfs = new LinkedList<>();
                    bfs.offer(new int[]{i, j, 0});
                    visited[i][j] = true;
                    
                    while (!bfs.isEmpty()) {
                        int[] current = bfs.poll();
                        int x = current[1];
                        int y = current[0];
                        int count = current[2];
                        
                        if (board[y].charAt(x) == 'G') {
                            return count;
                        }
                        
                        for (int k = 0; k < 4; k++) {
                            int xx = x;
                            int yy = y;
                            
                            while (xx >= 0 && xx < cols && yy >= 0 && yy < rows) {
                                if (board[yy].charAt(xx) == 'D') break;
                                xx += offset[k][1];
                                yy += offset[k][0];
                            }
                            
                            xx -= offset[k][1];
                            yy -= offset[k][0];
                            
                            if (visited[yy][xx]) continue;
                            visited[yy][xx] = true;
                            bfs.offer(new int[]{yy, xx, count + 1});
                        }
                    }
                    
                    break;
                }
            }
        }
        
        return -1;
    }
}

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.cpp
#include <bits/stdc++.h>
using namespace std;

int solution(vector<string> board) {
    int offset[4][2] = { {0,1},{0,-1},{1,0},{-1,0} };
    vector<vector<bool>> visited(board.size(), vector<bool>(board[0].size()));
    for (int i = 0; i < board.size(); i++) {
        for (int j = 0; j < board[i].size(); j++) {
            if (board[i][j] == 'R') {
                queue<pair<pair<int, int>, int>> bfs;
                bfs.push({ { i, j }, 0 });
                visited[i][j] = true;
                while (!bfs.empty()) {
                    int x = bfs.front().first.second;
                    int y = bfs.front().first.first;
                    int count = bfs.front().second;
                    bfs.pop();
                    if (board[y][x] == 'G')return count;
                    for (int k = 0; k < 4; k++) {
                        int xx = x;
                        int yy = y;
                        while (xx < board[i].size() && xx >= 0 && yy < board.size() && yy >= 0) {
                            if (board[yy][xx] == 'D')break;
                            xx += offset[k][1];
                            yy += offset[k][0];
                        }
                        xx -= offset[k][1];
                        yy -= offset[k][0];
                        if (visited[yy][xx])continue;
                        visited[yy][xx] = true;
                        bfs.push({ { yy, xx }, count + 1 });

                    }
                }
                break;
            }
        }
    }
    return -1;
}